File: java/ConsumerComplaints.java
package io.quicktype;

import java.util.Map;
import com.fasterxml.jackson.annotation.*;

public class ConsumerComplaints {
    private String id;
    private String name;
    private long averageRating;
    private long createdAt;
    private String description;
    private DisplayType displayType;
    private long downloadCount;
    private boolean hideFromCatalog;
    private boolean hideFromDataJSON;
    private Long indexUpdatedAt;
    private boolean newBackend;
    private long numberOfComments;
    private long oid;
    private Provenance provenance;
    private boolean publicationAppendEnabled;
    private long publicationDate;
    private long publicationGroup;
    private PublicationStage publicationStage;
    private String rowClass;
    private long rowsUpdatedAt;
    private RowsUpdatedBy rowsUpdatedBy;
    private long tableID;
    private long totalTimesRated;
    private long viewCount;
    private long viewLastModified;
    private ViewType viewType;
    private Grant[] grants;
    private Metadata metadata;
    private Owner owner;
    private Ratings ratings;
    private Right[] rights;
    private TableAuthor tableAuthor;
    private ConsumerComplaintFlag[] flags;
    private Boolean moderationStatus;
    private String category;
    private String[] tags;
    private ModifyingViewUid modifyingViewUid;

    @JsonProperty("id")
    public String getId() { return id; }
    @JsonProperty("id")
    public void setId(String value) { this.id = value; }

    @JsonProperty("name")
    public String getName() { return name; }
    @JsonProperty("name")
    public void setName(String value) { this.name = value; }

    @JsonProperty("averageRating")
    public long getAverageRating() { return averageRating; }
    @JsonProperty("averageRating")
    public void setAverageRating(long value) { this.averageRating = value; }

    @JsonProperty("createdAt")
    public long getCreatedAt() { return createdAt; }
    @JsonProperty("createdAt")
    public void setCreatedAt(long value) { this.createdAt = value; }

    @JsonProperty("description")
    public String getDescription() { return description; }
    @JsonProperty("description")
    public void setDescription(String value) { this.description = value; }

    @JsonProperty("displayType")
    public DisplayType getDisplayType() { return displayType; }
    @JsonProperty("displayType")
    public void setDisplayType(DisplayType value) { this.displayType = value; }

    @JsonProperty("downloadCount")
    public long getDownloadCount() { return downloadCount; }
    @JsonProperty("downloadCount")
    public void setDownloadCount(long value) { this.downloadCount = value; }

    @JsonProperty("hideFromCatalog")
    public boolean getHideFromCatalog() { return hideFromCatalog; }
    @JsonProperty("hideFromCatalog")
    public void setHideFromCatalog(boolean value) { this.hideFromCatalog = value; }

    @JsonProperty("hideFromDataJson")
    public boolean getHideFromDataJSON() { return hideFromDataJSON; }
    @JsonProperty("hideFromDataJson")
    public void setHideFromDataJSON(boolean value) { this.hideFromDataJSON = value; }

    @JsonProperty("indexUpdatedAt")
    public Long getIndexUpdatedAt() { return indexUpdatedAt; }
    @JsonProperty("indexUpdatedAt")
    public void setIndexUpdatedAt(Long value) { this.indexUpdatedAt = value; }

    @JsonProperty("newBackend")
    public boolean getNewBackend() { return newBackend; }
    @JsonProperty("newBackend")
    public void setNewBackend(boolean value) { this.newBackend = value; }

    @JsonProperty("numberOfComments")
    public long getNumberOfComments() { return numberOfComments; }
    @JsonProperty("numberOfComments")
    public void setNumberOfComments(long value) { this.numberOfComments = value; }

    @JsonProperty("oid")
    public long getOid() { return oid; }
    @JsonProperty("oid")
    public void setOid(long value) { this.oid = value; }

    @JsonProperty("provenance")
    public Provenance getProvenance() { return provenance; }
    @JsonProperty("provenance")
    public void setProvenance(Provenance value) { this.provenance = value; }

    @JsonProperty("publicationAppendEnabled")
    public boolean getPublicationAppendEnabled() { return publicationAppendEnabled; }
    @JsonProperty("publicationAppendEnabled")
    public void setPublicationAppendEnabled(boolean value) { this.publicationAppendEnabled = value; }

    @JsonProperty("publicationDate")
    public long getPublicationDate() { return publicationDate; }
    @JsonProperty("publicationDate")
    public void setPublicationDate(long value) { this.publicationDate = value; }

    @JsonProperty("publicationGroup")
    public long getPublicationGroup() { return publicationGroup; }
    @JsonProperty("publicationGroup")
    public void setPublicationGroup(long value) { this.publicationGroup = value; }

    @JsonProperty("publicationStage")
    public PublicationStage getPublicationStage() { return publicationStage; }
    @JsonProperty("publicationStage")
    public void setPublicationStage(PublicationStage value) { this.publicationStage = value; }

    @JsonProperty("rowClass")
    public String getRowClass() { return rowClass; }
    @JsonProperty("rowClass")
    public void setRowClass(String value) { this.rowClass = value; }

    @JsonProperty("rowsUpdatedAt")
    public long getRowsUpdatedAt() { return rowsUpdatedAt; }
    @JsonProperty("rowsUpdatedAt")
    public void setRowsUpdatedAt(long value) { this.rowsUpdatedAt = value; }

    @JsonProperty("rowsUpdatedBy")
    public RowsUpdatedBy getRowsUpdatedBy() { return rowsUpdatedBy; }
    @JsonProperty("rowsUpdatedBy")
    public void setRowsUpdatedBy(RowsUpdatedBy value) { this.rowsUpdatedBy = value; }

    @JsonProperty("tableId")
    public long getTableID() { return tableID; }
    @JsonProperty("tableId")
    public void setTableID(long value) { this.tableID = value; }

    @JsonProperty("totalTimesRated")
    public long getTotalTimesRated() { return totalTimesRated; }
    @JsonProperty("totalTimesRated")
    public void setTotalTimesRated(long value) { this.totalTimesRated = value; }

    @JsonProperty("viewCount")
    public long getViewCount() { return viewCount; }
    @JsonProperty("viewCount")
    public void setViewCount(long value) { this.viewCount = value; }

    @JsonProperty("viewLastModified")
    public long getViewLastModified() { return viewLastModified; }
    @JsonProperty("viewLastModified")
    public void setViewLastModified(long value) { this.viewLastModified = value; }

    @JsonProperty("viewType")
    public ViewType getViewType() { return viewType; }
    @JsonProperty("viewType")
    public void setViewType(ViewType value) { this.viewType = value; }

    @JsonProperty("grants")
    public Grant[] getGrants() { return grants; }
    @JsonProperty("grants")
    public void setGrants(Grant[] value) { this.grants = value; }

    @JsonProperty("metadata")
    public Metadata getMetadata() { return metadata; }
    @JsonProperty("metadata")
    public void setMetadata(Metadata value) { this.metadata = value; }

    @JsonProperty("owner")
    public Owner getOwner() { return owner; }
    @JsonProperty("owner")
    public void setOwner(Owner value) { this.owner = value; }

    @JsonProperty("ratings")
    public Ratings getRatings() { return ratings; }
    @JsonProperty("ratings")
    public void setRatings(Ratings value) { this.ratings = value; }

    @JsonProperty("rights")
    public Right[] getRights() { return rights; }
    @JsonProperty("rights")
    public void setRights(Right[] value) { this.rights = value; }

    @JsonProperty("tableAuthor")
    public TableAuthor getTableAuthor() { return tableAuthor; }
    @JsonProperty("tableAuthor")
    public void setTableAuthor(TableAuthor value) { this.tableAuthor = value; }

    @JsonProperty("flags")
    public ConsumerComplaintFlag[] getFlags() { return flags; }
    @JsonProperty("flags")
    public void setFlags(ConsumerComplaintFlag[] value) { this.flags = value; }

    @JsonProperty("moderationStatus")
    public Boolean getModerationStatus() { return moderationStatus; }
    @JsonProperty("moderationStatus")
    public void setModerationStatus(Boolean value) { this.moderationStatus = value; }

    @JsonProperty("category")
    public String getCategory() { return category; }
    @JsonProperty("category")
    public void setCategory(String value) { this.category = value; }

    @JsonProperty("tags")
    public String[] getTags() { return tags; }
    @JsonProperty("tags")
    public void setTags(String[] value) { this.tags = value; }

    @JsonProperty("modifyingViewUid")
    public ModifyingViewUid getModifyingViewUid() { return modifyingViewUid; }
    @JsonProperty("modifyingViewUid")
    public void setModifyingViewUid(ModifyingViewUid value) { this.modifyingViewUid = value; }
}

File: javascript/consumer-complaints.js
// To parse this data:
//
//   const Convert = require("./file");
//
//   const consumerComplaints = Convert.toConsumerComplaints(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
function toConsumerComplaints(json) {
    return cast(JSON.parse(json), a(o("ConsumerComplaints")));
}

function consumerComplaintsToJson(value) {
    return JSON.stringify(value, null, 2);
}

function cast(obj, typ) {
    if (!isValid(typ, obj)) {
        throw `Invalid value`;
    }
    return obj;
}

function isValid(typ, val) {
    if (typ === undefined) return true;
    if (typ === null) return val === null || val === undefined;
    return typ.isUnion  ? isValidUnion(typ.typs, val)
            : typ.isArray  ? isValidArray(typ.typ, val)
            : typ.isMap    ? isValidMap(typ.typ, val)
            : typ.isEnum   ? isValidEnum(typ.name, val)
            : typ.isObject ? isValidObject(typ.cls, val)
            :                isValidPrimitive(typ, val);
}

function isValidPrimitive(typ, val) {
    return typeof typ === typeof val;
}

function isValidUnion(typs, val) {
    // val must validate against one typ in typs
    return typs.find(typ => isValid(typ, val)) !== undefined;
}

function isValidEnum(enumName, val) {
    const cases = typeMap[enumName];
    return cases.indexOf(val) !== -1;
}

function isValidArray(typ, val) {
    // val must be an array with no invalid elements
    return Array.isArray(val) && val.every(element => {
        return isValid(typ, element);
    });
}

function isValidMap(typ, val) {
    if (val === null || typeof val !== "object" || Array.isArray(val)) return false;
    // all values in the map must be typ
    return Object.keys(val).every(prop => {
        if (!Object.prototype.hasOwnProperty.call(val, prop)) return true;
        return isValid(typ, val[prop]);
    });
}

function isValidObject(className, val) {
    if (val === null || typeof val !== "object" || Array.isArray(val)) return false;
    let typeRep = typeMap[className];
    return Object.keys(typeRep).every(prop => {
        if (!Object.prototype.hasOwnProperty.call(typeRep, prop)) return true;
        return isValid(typeRep[prop], val[prop]);
    });
}

function a(typ) {
    return { typ, isArray: true };
}

function e(name) {
    return { name, isEnum: true };
}

function u(...typs) {
    return { typs, isUnion: true };
}

function m(typ) {
    return { typ, isMap: true };
}

function o(className) {
    return { cls: className, isObject: true };
}

const typeMap = {
    "ConsumerComplaints": {
        id: "",
        name: "",
        averageRating: 0,
        createdAt: 0,
        description: "",
        displayType: e("DisplayType"),
        downloadCount: 0,
        hideFromCatalog: false,
        hideFromDataJson: false,
        indexUpdatedAt: u(null, 0),
        newBackend: false,
        numberOfComments: 0,
        oid: 0,
        provenance: e("Provenance"),
        publicationAppendEnabled: false,
        publicationDate: 0,
        publicationGroup: 0,
        publicationStage: e("PublicationStage"),
        rowClass: u(null, ""),
        rowsUpdatedAt: 0,
        rowsUpdatedBy: e("RowsUpdatedBy"),
        tableId: 0,
        totalTimesRated: 0,
        viewCount: 0,
        viewLastModified: 0,
        viewType: e("ViewType"),
        grants: a(o("Grant")),
        metadata: o("Metadata"),
        owner: o("Owner"),
        ratings: u(null, o("Ratings")),
        rights: a(e("Right")),
        tableAuthor: o("TableAuthor"),
        flags: u(null, a(e("ConsumerComplaintFlag"))),
        moderationStatus: u(null, false),
        category: u(null, ""),
        tags: u(null, a("")),
        modifyingViewUid: u(null, e("ModifyingViewUid")),
    },
    "Grant": {
        inherited: false,
        type: e("GrantType"),
        flags: a(e("GrantFlag")),
    },
    "Metadata": {
        jsonQuery: u(null, o("JSONQuery")),
        rdfSubject: u(null, e("RDFSubject")),
        rdfClass: u(null, ""),
        rowIdentifier: u(null, ""),
        availableDisplayTypes: a(e("AvailableDisplayType")),
        rowLabel: u(null, ""),
        renderTypeConfig: o("RenderTypeConfig"),
        richRendererConfigs: u(null, o("RichRendererConfigs")),
        custom_fields: u(null, o("CustomFields")),
    },
    "CustomFields": {
        TEST: o("Test"),
    },
    "Test": {
        CFPB1: "",
    },
    "JSONQuery": {
        order: u(null, a(o("Order"))),
        select: u(null, a(o("Select"))),
        group: u(null, a(o("Group"))),
    },
    "Group": {
        columnFieldName: "",
    },
    "Order": {
        ascending: false,
        columnFieldName: "",
    },
    "Select": {
        columnFieldName: "",
        aggregate: u(null, ""),
    },
    "RenderTypeConfig": {
        visible: o("Visible"),
    },
    "Visible": {
        table: u(null, false),
        fatrow: u(null, false),
    },
    "RichRendererConfigs": {
        fatRow: o("FatRow"),
    },
    "FatRow": {
        columns: a(o("Column")),
    },
    "Column": {
        styles: o("Styles"),
        rows: a(o("Row")),
    },
    "Row": {
        fields: a(o("Field")),
    },
    "Field": {
        tableColumnId: 0,
        type: e("FieldType"),
    },
    "Styles": {
        width: e("Width"),
    },
    "Owner": {
        id: "",
        displayName: "",
        screenName: "",
        type: e("OwnerType"),
        flags: u(null, a(e("OwnerFlag"))),
        profileImageUrlLarge: u(null, ""),
        profileImageUrlMedium: u(null, ""),
        profileImageUrlSmall: u(null, ""),
    },
    "Ratings": {
        rating: 0,
    },
    "TableAuthor": {
        id: e("ID"),
        displayName: e("Name"),
        screenName: e("Name"),
        type: e("OwnerType"),
        flags: u(null, a(e("OwnerFlag"))),
    },
    "DisplayType": [
        "fatrow",
        "table",
    ],
    "ConsumerComplaintFlag": [
        "default",
        "restorable",
        "restorePossibleForType",
    ],
    "GrantFlag": [
        "public",
    ],
    "GrantType": [
        "viewer",
    ],
    "AvailableDisplayType": [
        "fatrow",
        "page",
        "table",
    ],
    "RDFSubject": [
        "0",
    ],
    "FieldType": [
        "columnData",
        "columnLabel",
    ],
    "Width": [
        "27%",
        "30%",
        "31%",
        "33%",
        "40%",
    ],
    "ModifyingViewUid": [
        "s6ew-h6mp",
    ],
    "OwnerFlag": [
        "organizationMember",
    ],
    "OwnerType": [
        "interactive",
    ],
    "Provenance": [
        "official",
    ],
    "PublicationStage": [
        "published",
    ],
    "Right": [
        "read",
    ],
    "RowsUpdatedBy": [
        "pjxg-ve4m",
        "54a3-qyun",
        "9e3m-2843",
        "vvca-fr6g",
    ],
    "Name": [
        "Dan Munz",
        "Doug Taylor",
        "ming",
        "Shannon Files",
    ],
    "ID": [
        "pjxg-ve4m",
        "54a3-qyun",
        "9e3m-2843",
        "9eki-h2hn",
    ],
    "ViewType": [
        "tabular",
    ],
};

module.exports = {
    "consumerComplaintsToJson": consumerComplaintsToJson,
    "toConsumerComplaints": toConsumerComplaints,
};
